tron/eventbus.py

```diff
@@ -0,0 +1,162 @@
+        started = time.time()
+        new_file = os.path.join(self.log_dir, f"{int(started)}.pickle")
+        with open(new_file, 'xb') as f:
+            pickle.dump(self.event_log, f)
+
+        tmplink = os.path.join(self.log_dir, "tmp")
+        os.symlink(new_file, tmplink)
+        os.rename(tmplink, self.log_current)
+
+        self.log_last_save = time.time()
+        duration = self.log_last_save - started
+        log.info(f"log dumped to disk because {reason}, took {duration:.4}s")
+
+    def sync_loop(self):
+        if self.must_shutdown:
+            self.sync_save_log("shutdown")
+            log.info("shutdown completed")
+        else:
+            self.sync_process()
+            reactor.callLater(1, self.sync_loop)
+
+    def sync_process(self):
+        if time.time() > self.log_last_save + self.log_save_interval:
+            save_reason = f"{self.log_save_interval}s passed"
+        elif self.log_updates > self.log_save_updates:
+            save_reason = f"{self.log_save_updates} updates"
+        else:
+            save_reason = None
+
+        if save_reason:
+            self.sync_save_log(save_reason)
+
+        consume_dequeue(self.subscribe_queue, self.sync_subscribe)
+        consume_dequeue(self.publish_queue, self.sync_publish)
+
+    def sync_publish(self, event):
+        event = pickle.loads(pickle.dumps(event))
+        event_id = event['id']
+        if event_id in self.event_log:
+            if self.event_log[event_id] != event:
+                log.info(f"replacing event: {event_id}")
+            else:
+                log.debug(f"duplicate event: {event}")
+                return
+
+        self.event_log[event['id']] = event
+        self.log_save_updates += 1
+        log.debug(f"event stored: {event}")
+
+        reactor.callLater(0, self.sync_notify, event['id'])
+
+    def sync_subscribe(self, prefix_subscriber_cb):
+        prefix, subscriber, cb = prefix_subscriber_cb
+
+        if prefix in self.event_subscribers:
+            self.event_subscribers[prefix].append((subscriber, cb))
+        else:
+            self.event_subscribers[prefix] = [(subscriber, cb)]
+
+        log.debug(f"subscriber registered: {prefix_subscriber_cb}")
+
+    def sync_unsubscribe(self, prefix_sub):
+        prefix, sub = prefix_sub
+
+        if prefix not in self.event_subscribers:
+            log.debug(f"subscription  not found for prefix {prefix}")
+            return
+
+        new_subs = [
+            sub_cb for sub_cb in self.event_subscribers[prefix]
+            if sub_cb[0] != sub
+        ]
+        if new_subs:
+            self.event_subscribers[prefix] = new_subs
+        else:
+            del self.event_subscribers[prefix]
+        log.debug(f"subscription removed: {prefix} / {sub}")
+
+    def sync_notify(self, event_id):
+        event = self.event_log[event_id]
+        for prefix, subscribers in self.event_subscribers.items():
+            if event_id.startswith(prefix):
```

Comment:
This startswith seems funny to me. Does it need to be that loose?

Reply:
it doesn't need to be, but why not?